presentation-compiler/src/main/scala/org/alephium/ralph/lsp/pc/state/PCStateDiagnostics.scala

```diff
@@ -229,25 +229,44 @@ object PCStateDiagnostics {
   }
 
   /** Fetch workspace/project level diagnostics i.e. diagnostics that do have source information. */
-  def toWorkspaceDiagnostics(workspace: WorkspaceState.IsSourceAware): FileDiagnostic = {
-    val workspaceDiagnostics =
-      workspace match {
-        case compiled: WorkspaceState.Errored =>
-          compiled.workspaceErrors map {
-            error =>
-              // These are workspace level errors such as `Compiler.Error`, their source-code information is unknown.
-              toDiagnostic(
+  def toWorkspaceDiagnostics(workspace: WorkspaceState.IsSourceAware): Iterable[FileDiagnostic] = {
+    workspace match {
+      case compiled: WorkspaceState.Errored =>
+        // Group by fileURI provided errors
+        compiled.workspaceErrors groupBy(_.index.fileURI) map {
+          case (Some(fileURI), errors) =>
+            // fileURI is known, we try to find the source code.
+            compiled.sourceCode.find(_.fileURI == fileURI) match {
+              case Some(source) =>
+                // Source code is known, we report errors at source-code level.
```

Comment:
This is great! Just a minor adjustment: we can move this logic to execute immediately after we [compile the contracts](https://github.com/alephium/ralph-lsp/blob/00fcd7a0080aff63633f113df1e7a5878199ba9a/presentation-compiler/src/main/scala/org/alephium/ralph/lsp/pc/sourcecode/SourceCode.scala#L270). This way, compilation errors containing `fileURI` won't be transformed into `workspaceErrors` at all. Instead, it will be recognised throughout the code, rather than only when responding to the client in diagnostics.

I've submitted PR #130 to this branch with the change.